### ISSUE TITLE
Font rendering of ligatures (and possibly other features?) is sometimes inconsistent

### ISSUE THREAD

**Issue by nerdo** [wez/wezterm#4874]
### What Operating System(s) are you seeing this problem on?

macOS

### Which Wayland compositor or X11 Window manager(s) are you using?

macOS Sonoma 14.2.1 (23C71)

### WezTerm version

20240127-090133-2a8bdfa6

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

I mentioned this in another issue because I thought I was having the same issue originally but it seems like a separate issue altogether.

I have encountered some odd font rendering issues. When and where they happen seems to be very inconsistent so it's a little bit difficult to describe in words, but for the most part it seems like rendering ligatures is inconsistent.

Variations I've encountered:
- Parts of a ligature appear correct, but the entire thing seems to occupy the wrong horizontal or vertical space
- Fonts sometimes shift vertically slightly out of the blue
- Underscores sometimes disappear (might be a symptom of the above)

Any and/or all of these might occur in different places in the same screen.

Here's a screenshot from a neovim editing session to illustrate:

![Screenshot 2024-01-27 at 6 57 56 AM](https://github.com/wez/wezterm/assets/1031502/950ea09c-d5db-437f-82e1-a00ddd3d325d)

Notice the `->`'s on the lines where `$factory->define` is.

The first line renders correctly. I have a freely available font `Caskaydia Cove Nerd Font Mono` which has a ligature for `->` to turn it into a connected arrow.

However, if you look at the other two instances of the same exact text, they are rendered incorrrectly. It appears as if the ligature is there, but it's not taking up the same space.

Also notice all of the fat arrows `=>` are also ligatures, but they two characters making it up don't line up properly so there is a stair step effect.

_Originally posted by @nerdo in https://github.com/wez/wezterm/issues/3774#issuecomment-1913138586_

### To Reproduce

I don't really know how to reproduce it.

It seems to happen randomly.

### Configuration

local wezterm = require("wezterm")
return {
	disable_default_key_bindings = true,
	adjust_window_size_when_changing_font_size = false,
	enable_tab_bar = false,
	front_end = "OpenGL",
	font = wezterm.font("CaskaydiaCove Nerd Font Mono"),
	font_size = 16,
	line_height = 1.6,
	color_scheme = "Argonaut",
	colors = {
		cursor_bg = "#800500",
		cursor_fg = "#eacecd",
	},
	-- make both left and right alt behave in the same, normal, way
	-- https://wezfurlong.org/wezterm/config/keyboard-concepts.html?highlight=right%20option#macos-left-and-right-option-key
	send_composed_key_when_left_alt_is_pressed = false,
	send_composed_key_when_right_alt_is_pressed = false,
	keys = {
		{ key = "v", mods = "CTRL|SHIFT", action = wezterm.action({ PasteFrom = "Clipboard" }) },
		{ key = "=", mods = "CTRL", action = "IncreaseFontSize" },
		{ key = "-", mods = "CTRL", action = "DecreaseFontSize" },
		{ key = "0", mods = "CTRL", action = "ResetFontSize" },

		-- MacOS keybindings
		{ key = "c", mods = "CMD", action = wezterm.action.CopyTo("ClipboardAndPrimarySelection") },
		{ key = "v", mods = "CMD", action = wezterm.action.PasteFrom("Clipboard") },
		{ key = "m", mods = "CMD", action = "Hide" },
		{ key = "q", mods = "CMD", action = "QuitApplication" },
		{ key = "0", mods = "CMD", action = "ResetFontSize" },
		{ key = "=", mods = "CMD", action = "IncreaseFontSize" },
		{ key = "-", mods = "CMD", action = "DecreaseFontSize" },
	},
}

### Anything else?

I've tried tinkering with various font and rendering settings and various different fonts. Sometimes the problems appear to go away for a little while but any workarounds I've tried so far have been red herrings, as the issues always seem to recur later.

Prior to writing this, I was using the stable version of wezterm and updated to the nightly.

At first glance, everything appeared to be working great, but when I scrutinized the same screen I had open in the screenshot, I found different artifacts being expressed this time:

![Screenshot 2024-01-27 at 1 22 57 PM](https://github.com/wez/wezterm/assets/1031502/88152cb5-ad5b-4c61-8372-9ce421f2f5b8)

Note the text `$faker` on the highlighted line towards the bottom that reads `$factory->state(App\Concert::class, 'published', function ($faker) {`. This appears to render correctly here, but in the next function, and even the first function above, there are several artifacts on the letter `a` where it seems like it is vertically pushed down a pixel or so making it appear as if the bottom of the letter is cut off.

I think this artifact expresses itself in other areas of this screen, like the `f` in faker, too, but it's hard to tell because it's so subtle on letters where there's no clear bottom.

**Comment by wez** [comment]
Try setting:

```lua
config.freetype_load_flags = 'NO_HINTING'
```

**Comment by nerdo** [comment]
At first glance, like before (see https://github.com/wez/wezterm/issues/3774#issuecomment-1852663247 and https://github.com/wez/wezterm/issues/3774#issuecomment-1857163595) it *does* seem to correct it.

I'll give it a few days and see if I notice any more issues.

**Comment by wez** [comment]
I've made `NO_HINTING` the default in current nightly builds in 616b218e328dd47e859beddc251c1980c49358a9

**Comment by nerdo** [comment]
Thanks wez.

Ligatures, so far, appear better, but today, I'm seeing weird vertical rendering of the letter `s` today.

This is often how the problem manifests btw. Today it's this character, tomorrow, it's something else. Or even from screen to screen things can sometimes shift, so I don't think I always catch it happening right away.

![Screenshot 2024-01-29 at 9 10 43 AM](https://github.com/wez/wezterm/assets/1031502/6146aa66-4856-464d-ba60-02c1867cde71)

Notice how the `s` on the line `function user_can_view_a_published_concert_listing()` looks a bit... vertically squashed?

Then a few lines down, it appears to have the correct height, but its baseline shifted down a pixel or two... actually, yeah, in the word `states` on the line `        $concert =  factory(Concert::class)->states('published')->create([`, it looks like the two `s`'s are rendered slightly differently.

The same artifact shows up on `        $this->visit('/concerts/' . $concert->id);`

This might be a separate issue but it seems related. I can open a new issue if you think that's best?

FYI After reading your last comment, and before posting this, I updated to `wezterm 20240128-202157-1e552d76` and kept my config the same as shown [here](https://github.com/nerdo/personal-settings/commit/f3de6ec807d197357eee9dc24973220426755f31).

**Comment by rgruyters** [comment]
I have still issues with it with the latest nightly version, `20240130-221529-1479c7be aarch64-apple-darwin`.

I have this issue only when using Monaspace font with harfbuzz_features set. Without the harfbuzz_features it looks fine.

(with harfbuzz_features set)
![Screenshot 2024-01-30 at 10 38 56](https://github.com/wez/wezterm/assets/2082795/13636e0b-5ef4-4d3c-88c1-a67efda9fa27)

**Comment by 7moodDxB2004** [comment]
Hi all, I had the same issue on windows, I use fullscreen mode which might have an effect. The solution for me was to experiment with the font_size and line_height, and setting them to fractional numbers, here's an example for the difference between font_size 12.0 (1st image), 12.8 (2nd image), and 13.0 (3rd image), all while line_height is 1.1:

![Screenshot 2024-03-27 145724](https://github.com/wez/wezterm/assets/138982423/11b85648-1e93-4b48-b759-774eb2a6cf76)
![Screenshot 2024-03-27 151102](https://github.com/wez/wezterm/assets/138982423/939a95b6-38bb-452f-8844-c8437063c4e0)
![Screenshot 2024-03-27 145810](https://github.com/wez/wezterm/assets/138982423/6635269e-d112-4a54-8d9d-b9213696fe6f)

Note: you can see, when font_size is set to 12.8, all of these characters are aligned, this is really noticeable in line 21 `window_decorations = 'INTEGRATED_BUTTONS|RESIZE',`

if you want a change that will have a minimal impact on the appearance, then change line_height, I had it at 1.1, font_size at 12.0 and had this issue, then changed it to 1.08, and it was fixed, hope this helps.

**Comment by 7moodDxB2004** [comment]
You're right about JetBrains Mono, I use the nerd font version I installed on windows, normally this problem doesn't exist, I think when I lowered the `font_size` and `line_height` in wezterm, this issue started, even with the normal JetBrains Mono.

As for the minimum, I don't think there's a minimum, I tried `font_size = 0.02` and `line_height = 0.02` and it still worked, but characters were dots, if that's what you mean.

I think when you see this issue, it means the `line_height` is more than the height of the row and from my understanding, rows and columns can't have a fractional value.

**Comment by my4ng** [comment]
I have experienced a similar issue with Caskaydia Cove at font size 16.0 (and some larger size, e.g. 19, 20, 21, 22):
![image](https://github.com/wez/wezterm/assets/17978089/5f051d9e-20c5-408f-a9b0-db51e0849fcb)
Setting `freetype_load_flags = "NO_HINTING"` improves it, but there are still some artifacts ('bumps' between characters) and does appear to be more 'blurry':
![image](https://github.com/wez/wezterm/assets/17978089/8141e1e7-8b85-49b7-a693-50fb1114f7af)
My current workaround is tweaking the font size to 16.5 with hinting:
![image](https://github.com/wez/wezterm/assets/17978089/b1825f3d-edb4-4534-8da9-15a76a3b9e41)
